### PR TITLE
Issue #1030: InvalidationValve

### PR DESCRIPTION
**Pull request by ljacomet**
* InvalidationValve interface and wiring
* Tests
* Delete unused code

### CODE REVIEW

File: core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java [hunk 1]
```diff
@@ -64,6 +64,33 @@
...
+   */
+  void setInvalidationValve(InvalidationValve<K> valve);
+
+  /**
+   * Invalidation valve, that is the mechanism through which an {@link AuthoritativeTier} can request invalidations
+   * from the {@link CachingTier}.
+   *
+   * @param <K> the key type
+   */
+  interface InvalidationValve<K> {
```
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java]: The generic type is not needed.

File: core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java [hunk 2]
```diff
@@ -64,6 +64,33 @@
   boolean flush(K key, ValueHolder<V> valueHolder);
 
   /**
+   * Injects the {@link InvalidationValve} to be used by this {@code AuthoritativeTier}.
```
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java]: Sets... the caller is injecting... the receiver is just setting.

File: core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java [hunk 3]
```diff
@@ -64,6 +64,33 @@
...
+   * <P>
+   *   The role of the invalidation valve is to provide a way for the {@code AuthoritativeTier} to force invalidation of
```
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java]: passive voice... "The invalidation valve provides a way..."

File: core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java [hunk 4]
```diff
@@ -64,6 +64,33 @@
...
+  /**
+   * Invalidation valve, that is the mechanism through which an {@link AuthoritativeTier} can request invalidations
+   * from the {@link CachingTier}.
+   *
+   * @param <K> the key type
+   */
+  interface InvalidationValve<K> {
+
+    /**
+     * Requests an invalidation of all {@link CachingTier} mappings.
```
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java]: Is there a reason this is "all mappings"?  Would "some" mappings suffice?  Do we call this thing in some kind of loop or not?

File: impl/src/main/java/org/ehcache/impl/internal/store/offheap/AbstractOffHeapStore.java [hunk 5]
```diff
@@ -1115,10 +1120,6 @@ private void notifyInvalidation(final K key, final ValueHolder<V> p) {
     }
   }
 
-  public void registerEmergencyValve(final Callable<Void> valve) {
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/impl/internal/store/offheap/AbstractOffHeapStore.java]: Any reason for rejecting the old name?

File: impl/src/main/java/org/ehcache/impl/internal/store/offheap/AbstractOffHeapStore.java [hunk 6]
```diff
@@ -1200,11 +1201,11 @@ public Boolean apply() {
     }
   }
 
-  public void handleOversizeMappingException(K key, OversizeMappingException cause, StoreEventSink<K, V> eventSink) throws StoreAccessException {
+  void handleOversizeMappingException(K key, OversizeMappingException cause, StoreEventSink<K, V> eventSink) throws StoreAccessException {
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/impl/internal/store/offheap/AbstractOffHeapStore.java]: unrelated?

File: core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java (hunk 1 above)
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java]: Right, made it there when I was trying to add the `invalidate(K key)` as well. Will remove for now.

File: core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java (hunk 2 above)
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java]: Fixing

File: core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java (hunk 3 above)
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java]: Fixing

File: core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java (hunk 4 above)
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java]: How do we define _some_? In Ehcache 2.x this is actually called only once ... which makes me think, the current code most likely never triggers it 😨

File: impl/src/main/java/org/ehcache/impl/internal/store/offheap/AbstractOffHeapStore.java (hunk 5 above)
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/store/offheap/AbstractOffHeapStore.java]: Yes, I know this will become more in clustered

Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/impl/internal/store/offheap/AbstractOffHeapStore.java]: Fair enough.

File: impl/src/main/java/org/ehcache/impl/internal/store/offheap/AbstractOffHeapStore.java (hunk 6 above)
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/store/offheap/AbstractOffHeapStore.java]: Not really, that came with the Ehcache 2.x port and was never looked at. In this context, I looked at it and this public is wrong.

File: core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java (hunk 4 above)
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/core/spi/store/tiering/AuthoritativeTier.java]: Less than all?  Given what the invalidation valve is used for, evicting a single key might be good enough.  Of course if we want to log this maybe we should just invalidate everything and be done with it.

**Comment by ljacomet** [comment]
Adressed most of the comments.

This can be merged but the #1030 work is not complete yet - see comments over there.

File: impl/src/main/java/org/ehcache/impl/internal/store/offheap/AbstractOffHeapStore.java [hunk 7]
```diff
@@ -998,30 +1005,26 @@ public void invalidate(final K key) throws StoreAccessException {
...
-            notifyInvalidation(k, offHeapValueHolder);
-          }
-          function.apply();
-          return null;
+  public void invalidateAll() throws StoreAccessException {
+    invalidateAllObserver.begin();
+    StoreAccessException exception = null;
+    long errorCount = 0;
+    Iterator<Cache.Entry<K, ValueHolder<V>>> iterator = iterator();
+    while (iterator.hasNext()) {
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/impl/internal/store/offheap/AbstractOffHeapStore.java]: Any reason for not using a for loop here?

File: core-spi-test/src/main/java/org/ehcache/internal/tier/CachingTierGetOrComputeIfAbsent.java [hunk 8]
```diff
@@ -96,7 +96,7 @@ public void returnTheValueHolderCurrentlyInTheCachingTier() throws LegalSPITeste
     when(computedValueHolder.value()).thenReturn(value);
     when(computedValueHolder.expirationTime(any(TimeUnit.class))).thenReturn(Store.ValueHolder.NO_EXPIRE);
 
-    tier = factory.newCachingTier(1L);
+    tier = factory.newCachingTier();
```
Comment (chrisdennis) [inline core-spi-test/src/main/java/org/ehcache/internal/tier/CachingTierGetOrComputeIfAbsent.java]: Is this intentional?

File: core-spi-test/src/main/java/org/ehcache/internal/tier/CachingTierInvalidate.java [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+public class CachingTierInvalidate<K, V> extends CachingTierTester<K, V> {
+
+  private CachingTier<K, V> tier;
+
+  public CachingTierInvalidate(CachingTierFactory<K, V> factory) {
+    super(factory);
+  }
+
+  @Before
+  public void setUp() {
```
Comment (chrisdennis) [inline core-spi-test/src/main/java/org/ehcache/internal/tier/CachingTierInvalidate.java]: no-op method can be deleted?

File: impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java [hunk 10]
```diff
@@ -841,6 +834,59 @@ public void silentInvalidate(K key, final Function<Store.ValueHolder<V>, Void> f
...
+    invalidateAllObserver.end(CachingTierOperationOutcomes.InvalidateAllOutcome.SUCCESS);
+  }
+
+  @Override
+  public void silentInvalidateAll(final BiFunction<K, ValueHolder<V>, Void> biFunction) throws StoreAccessException {
+    silentInvalidateAllObserver.begin();
+    StoreAccessException exception = null;
+    long errorCount = 0;
+    Iterator<Cache.Entry<K, ValueHolder<V>>> iterator = iterator();
+    while (iterator.hasNext()) {
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java]: for loop?

File: core-spi-test/src/main/java/org/ehcache/internal/tier/CachingTierGetOrComputeIfAbsent.java (hunk 8 above)
Comment (ljacomet) [inline core-spi-test/src/main/java/org/ehcache/internal/tier/CachingTierGetOrComputeIfAbsent.java]: Yes, with some of the fixes around the bytesized on heap, this caused issues. And the size is not really a factor unless you test eviction.

File: impl/src/main/java/org/ehcache/impl/internal/store/offheap/AbstractOffHeapStore.java (hunk 7 above)
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/store/offheap/AbstractOffHeapStore.java]: well ... no :wink:

File: core-spi-test/src/main/java/org/ehcache/internal/tier/CachingTierInvalidate.java (hunk 9 above)
Comment (ljacomet) [inline core-spi-test/src/main/java/org/ehcache/internal/tier/CachingTierInvalidate.java]: Sure it can
